Lib/test/test_typing.py

```diff
@@ -7519,6 +7519,64 @@ class GenericNamedTuple(NamedTuple, Generic[T]):
 
         self.assertEqual(CallNamedTuple.__orig_bases__, (NamedTuple,))
 
+    def test_setname_called_on_values_in_class_dictionary(self):
+        class Vanilla:
+            def __set_name__(self, owner, name):
+                self.name = name
+
+        class Foo(NamedTuple):
+            attr = Vanilla()
+
+        foo = Foo()
+        self.assertEqual(len(foo), 0)
+        self.assertNotIn('attr', Foo._fields)
+        self.assertIsInstance(foo.attr, Vanilla)
+        self.assertEqual(foo.attr.name, "attr")
+
+        class Bar(NamedTuple):
+            attr: Vanilla = Vanilla()
+
+        bar = Bar()
+        self.assertEqual(len(bar), 1)
+        self.assertIn('attr', Bar._fields)
+        self.assertIsInstance(bar.attr, Vanilla)
+        self.assertEqual(bar.attr.name, "attr")
+
+    def test_setname_raises_the_same_as_on_other_classes(self):
+        class CustomException(BaseException): pass
+
+        class Annoying:
+            def __set_name__(self, owner, name):
+                raise CustomException("Cannot do that!")
+
+        with self.assertRaisesRegex(CustomException, "Cannot do that!") as cm:
+            class NormalClass:
+                attr = Annoying()
+        normal_exception = cm.exception
+
+        with self.assertRaisesRegex(CustomException, "Cannot do that!") as cm:
+            class NamedTupleClass:
```

Comment:
Did you mean to inherit this one from NamedTuple
```suggestion
            class NamedTupleClass(NamedTuple):
```

Reply:
oof. yes.